### PR TITLE
[FIXED JENKINS-28635] Show URL used by autoinstaller

### PR DESCRIPTION
**Pull request by vjuranek**
https://issues.jenkins-ci.org/browse/JENKINS-28635
Adds link on the right of the version select
jenkins-28635_1
when clicked, shows URL to use for the download
jenkins-28635_2

### CODE REVIEW

File: core/src/main/java/hudson/tools/DownloadFromUrlInstaller.java [hunk 1]
```diff
@@ -148,6 +146,16 @@ public String getId() {
...
             return Arrays.asList(((InstallableList)JSONObject.toBean(d,InstallableList.class)).list);
         }
+        
+        /**
+         * The UI uses this information to populate the drop-down.
+         * 
+         * @return JSON array of installable tools, never null
+         */
+        public JSONArray getInstallablesJson() throws IOException {
+            JSONObject d = Downloadable.get(getId()).getData();
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/tools/DownloadFromUrlInstaller.java]: <code>Downloadable.get()</code> may return null.
It is a mistake in the original code, but it makes sense to fix it.

**Comment by oleg-nenashev** [comment]
:+1: for the feature
Probably it makes sense to move the code to a taglib in order to reuse it in other code snippets using <code>Downloadable</code> (e.g. <code>j:downloadableSelector</code>)

File: core/src/main/resources/hudson/tools/DownloadFromUrlInstaller/config.jelly [hunk 2]
```diff
@@ -25,17 +25,41 @@ THE SOFTWARE.
...
+          function getDownloadURL(id, tools) {
+            for(var i = 0; i &lt; tools.length; i++) {
+              if(id === tools[i].id) {
+                return tools[i].url;
+              }
+            }
+          }
+          
+          function showDownloadURL(elem, tools) {
+            elem.nextSibling.childNodes[2].innerHTML = '&amp;nbsp;&amp;nbsp;Will be downloaded from ' + getDownloadURL(elem.value, tools);
```
Comment (olivergondza) [inline core/src/main/resources/hudson/tools/DownloadFromUrlInstaller/config.jelly]: This is not localizable, printing the url only should be self explanatory IMO.

File: core/src/main/java/hudson/tools/DownloadFromUrlInstaller.java [hunk 3]
```diff
@@ -148,6 +146,16 @@ public String getId() {
             if(d==null)     return Collections.emptyList();
             return Arrays.asList(((InstallableList)JSONObject.toBean(d,InstallableList.class)).list);
         }
+        
+        /**
+         * The UI uses this information to populate the drop-down.
+         * 
+         * @return JSON array of installable tools, never null
+         */
+        public JSONArray getInstallablesJson() throws IOException {
```
Comment (olivergondza) [inline core/src/main/java/hudson/tools/DownloadFromUrlInstaller.java]: Either `@since` or `@Restriced` for public method.

**Comment by olivergondza** [comment]
:+1: aside from the minor comments.

**Comment by olivergondza** [comment]
Personally, I do not think it is necessary hide the URL behind the "(Show download URL)" link.

**Comment by vjuranek** [comment]
thanks for review. I would keep hidden URL, as some URLs can be quite long and can clutter UI. Other comments will address late on in the code.

**Comment by vjuranek** [comment]
PR updated

File: core/src/main/java/hudson/tools/DownloadFromUrlInstaller.java [hunk 4]
```diff
@@ -148,6 +146,23 @@ public String getId() {
...
             return Arrays.asList(((InstallableList)JSONObject.toBean(d,InstallableList.class)).list);
         }
+        
+        /**
+         * The UI uses this information to populate the drop-down.
+         * 
+         * @return JSON array of installable tools, never null
+         * @since 1.617
+         */
+        public JSONArray getInstallablesJson() throws IOException {
```
Comment (daniel-beck) [inline core/src/main/java/hudson/tools/DownloadFromUrlInstaller.java]: `@Restricted(DoNotUse.class)`

Comment (vjuranek) [inline core/src/main/java/hudson/tools/DownloadFromUrlInstaller.java]: Could you elaborate, which it should be restricted? I don't see any issue with providing access to versions and URLs downloaded by back end crawlers and left this annotation intentionally. Thanks.

Comment (daniel-beck) [inline core/src/main/java/hudson/tools/DownloadFromUrlInstaller.java]: It's new API that only exists for the UI. If someone wants to use it, the annotation can always be removed later, but for now, it's just more to maintain for no reason.

Comment (vjuranek) [inline core/src/main/java/hudson/tools/DownloadFromUrlInstaller.java]: my understanding is, that `@Restricted` should be used only for deprecated API, which will be removed soon or API which is dangerous to use for any other reason (and IMHO in concordance with [javadoc](https://github.com/kohsuke/access-modifier/blob/master/access-modifier-annotation/src/main/java/org/kohsuke/accmod/Restricted.java#L34)), but if it helps this PR to be merged, I have no problem with it... PR updated

Comment (olivergondza) [inline core/src/main/java/hudson/tools/DownloadFromUrlInstaller.java]: I am not sure what the original intention was but the javadoc does indeed sound misleading for the use-cases we use it for.

Generally, we try to expose as little API as possible. And as stapler/jelly require public access modifier in several places, `@Restricted` is the second best thing. `DoNotUse` for what is accessed via reflection only, `NoExternalUse` for what needs to be referenced from the project only (tests usually).

**Comment by olivergondza** [comment]
:+1:

**Comment by oleg-nenashev** [comment]
:+1:

File: core/src/main/resources/lib/form/installableSelect.jelly [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    <j:otherwise>
+      <select name="_.id" onChange="showDownloadURL(this, ${tools})">
+        <j:forEach var="tool" items="${tools}">
+          <f:option value="${tool.id}" selected="${tool.id==instance.id}">${tool.name}
+          </f:option>
+        </j:forEach>
+      </select>
+      <span name="_.tool.url" style="color:gray">
+        &amp;nbsp;&amp;nbsp;
+        <a href="javascript:void(0)" onclick="makeUrlVisible(this, ${tools})">(Show download URL)</a>
```
Comment (daniel-beck) [inline core/src/main/resources/lib/form/installableSelect.jelly]: I assume this can be made `${%localizable}`?

**Comment by orrc** [comment]
I'm not sure this adds much value, and the URL could change between the time you select the installable item, and when it gets downloaded.  That is not super likely, but there are at least two tool installers broken just now that need their URLs changed.

As mentioned in the PR linked from JIRA, the actual download URL used is always logged to the console log when a build occurs with a tool installer.  It doesn't really provide much value to show the URL at global configuration time.

In any case, I don't believe this PR will work for a number of tool installers, e.g. I *think* the JDK and Go installers will fail (presumably throwing a `JSONException`), as this PR expects that there is a standardised JSON format for tool installers — a `list` array, with each object in the array having an `id` and `url`.
But that is not true, especially for installers that calculate the download URL at build time based on the type of machine it's running on.

**Comment by oleg-nenashev** [comment]
Cannot be merged due to the conflict

**Comment by oleg-nenashev** [comment]
@vjuranek 
Could you fix the merge conflict?

**Comment by oleg-nenashev** [comment]
Somebody needs to finalize this PR

File: core/src/main/resources/lib/form/installableSelect.jelly (hunk 5 above)
Comment (oleg-nenashev) [inline core/src/main/resources/lib/form/installableSelect.jelly]: It can

**Comment by orrc** [comment]
As per my comment, I don't think this feature is actually really useful.. it only really came up because a certain developer who likes to argue about irrelevant things did so.

Plus, as I mentioned, this code will anyway likely crash with certain tool installers. So it should either be closed, or still be marked as work-in-progress.

**Comment by oleg-nenashev** [comment]
> Plus, as I mentioned, this code will anyway likely crash with certain tool installers. So it should either be closed, or still be marked as work-in-progress.

We have "needs-fix", which indicates it is not going to fly in the current state. "work-in-progress" has been removed since it does not seem to be actually in progress. Now anybody can pick and finalize this PR if he wants

**Comment by orrc** [comment]
> We have "needs-fix"

Oops, sorry — was on mobile and missed the tags up at the top of the page.

**Comment by daniel-beck** [comment]
PR appears to have been abandoned, closing.

**Pull request closed without merging**